src/Database/LSMTree/Internal/IndexCompact.hs

```diff
@@ -436,42 +436,67 @@ pageSpanSize :: PageSpan -> NumPages
 pageSpanSize pspan = NumPages . toEnum $
     unPageNo (pageSpanEnd pspan) - unPageNo (pageSpanStart pspan) + 1
 
--- | Given a search key, find the number of the disk page that /might/ contain
--- this key.
+-- |  Searches for a page span that contains a key–value pair with the given key.
 --
--- Note: there is no guarantee that the disk page contains the key. However, it
--- is guaranteed that /no other/ disk page contain the key.
+-- If there is indeed such a pair, the result is the corresponding page span; if
+-- there is no such pair, the result is an arbitrary but valid page span.
 --
 -- See [an informal description of the search algorithm](#search-descr) for more
 -- details about the search algorithm.
---
--- TODO: optimisation ideas: we should be able to get it all unboxed,
--- non-allocating and without unnecessary bounds checking.
---
+search :: SerialisedKey -> IndexCompact -> PageSpan
 -- One way to think of the search algorithm is that it starts with the full page
 -- number interval, and shrinks it to a minimal interval that contains the
--- search key. The code below is annotated with @Pre:@ and @Post:@ comments that
--- describe the interval at that point.
-search :: SerialisedKey -> IndexCompact -> PageSpan
+-- search key. The code below is annotated with [x,y] or [x, y) comments that
+-- describe the known page number interval at that point in the search
+-- algorithm.
 search k IndexCompact{..} =
-    let !primbits = keyTopBits64 k
-    in
-      case unsafeSearchLE primbits icPrimary of
-        Nothing -> singlePage (PageNo 0)  -- Post: @[lb, lb)@ (empty).
-        Just !i ->         -- Post: @[lb, i]@.
-          if unBit $ icClashes VU.! i then
-            -- Post: @[lb, i]@, now in clash recovery mode.
-            let !i1  = PageNo $ fromMaybe 0 $
-                  bitIndexFromToRev (BoundInclusive 0) (BoundInclusive i) (Bit False) icClashes
-                !i2  = maybe (PageNo 0) snd $ Map.lookupLE (unsafeNoAssertMakeUnslicedKey k) icTieBreaker
-                PageNo !i3 = max i1 i2 -- Post: the intersection of @[i1, i]@ and @[i2, i].
-                !i4 = bitLongestPrefixFromTo (BoundExclusive i3) (BoundInclusive i) (Bit True) icLargerThanPage
-                      -- Post: [i3, i4]
-            in  if i3 == i4 then singlePage (PageNo i3) else multiPage (PageNo i3) (PageNo i4)
```

Comment:
INFO: I noticed that this is an unnecessary check. `multiPage (PageNo i3) (PageNo i4)` provides the same result